anvill/python/anvill/binja/bnprogram.py

```diff
@@ -75,9 +86,137 @@ def get_variable_impl(self, address):
         if isinstance(var_type, VoidType):
             var_type = ArrayType()
             var_type.set_num_elements(1)
+            return
 
         return BNVariable(bn_var, arch, address, var_type)
 
+    def _get_function_from_bnvariable(self, address, bn_var):
+        """Get the `Function` instance from the data variable of function
+        pointer type. Raise an `InvalidFunctionException` exception if
+        the data variable is not of function pointer type.
+        """
+        if bn_var is None:
+            return None
+
+        # if the bn_var is an external symbol; discard it It will not get
+        # recovered as function or variables
+        symbol = self._bv.get_symbol_at(address)
+        if symbol != None and symbol.type == bn.SymbolType.ExternalSymbol:
+            return None
+
+        if not _is_function_pointer(bn_var):
+            raise InvalidFunctionException(
+                "No function pointer is defined at address {:x}".format(address)
+            )
+
+        arch = self._arch
+        func_tinfo = bn_var.type.element_type
+
+        # TODO(akshayk): The type information may not have calling convention
+        # information. It does not recover the parameters and get the function
+        # in such case. A working solution could be handing the default calling
+        # convention for each architecture. This is in todo list.
+        if func_tinfo.calling_convention is None:
+            return None
+
+        calling_conv = CallingConvention(arch, func_tinfo)
+
+        func_type = self.type_cache.get(func_tinfo)
+
+        index = 0
+        param_list = []
+        stack_offset = arch.stack_offset()
+        for var in func_tinfo.parameters:
+            var_type = var.type
+            reg_name = None
+
+            if var_type is None:
+                continue
+
+            arg_type = self.type_cache.get(var_type)
+
+            if (
+                bn.TypeClass.IntegerTypeClass == var_type.type_class
+                or bn.TypeClass.PointerTypeClass == var_type.type_class
+            ):
+                reg_name = calling_conv.next_int_arg_reg
```

Comment:
What happens if we run out of integer argument registers? Should be be re-thinking this, perhaps letting our function specs either declare: returns+params, OR declare a single function type, then allow our ABI handling logic on the C++ side to do the work?

Reply:
Here we approximate the ABI while generating spec and follow what the calling convention says. There are instances in binja where the function type does not follow ABI. We discard arguments and force them to follow the ABI. It comes down to how reliable is decompiler in function type identification and may decide to follow that and let the C++ side handle the ABI logic.